### PR TITLE
RDRP-1087 outliers

### PR DESCRIPTION
**Pull request by AnneONS**
## Pull Request submission

The code src.outlier_detection.auto_outliers.py has a large function flag_outliers. 

I have created a separate function that indicates the number or rows that should be clipped for outliering, with a unit test.

This unit test illustrates how many rows get outliered depending on the size of valid entries in a cell.

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #1087

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [x] All requirements install from (updated) `requirements.txt`
- [x] Documentation has been created and is clear - **check the working document**
- [x] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [x] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [x] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/outlier_detection/auto_outliers.py [hunk 1]
```diff
@@ -95,32 +119,32 @@ def flag_outliers(
...
-        filtered_df["group_rank"] <= filtered_df["lower_band"]
-    )
+    # Calculate the number of rows to clip by merging in the group count
+    filtered_df = filtered_df.merge(count_df, how="left", on=groupby_cols)
+
+    # Create outlier condition
+    outlier_cond = filtered_df["group_rank"] > filtered_df["clip_higher_than"]
+
+    # If lower clipping is specified, add the condition
+    if lower_clip > 0:
```
Comment (adriano-lopresti) [inline src/outlier_detection/auto_outliers.py:140]: the case where function "get_clip_bands" has a non-zero lower_clip has not been tested. I'm wondering if this is worth doing.
Also, has "flag_outliers" been unit-tested for the case where lower_clip is not zero?

**Review by adriano-lopresti — CHANGES_REQUESTED** [review]
When running the pipeline I get the following error:

FileExistsError: File: R:/BERD Results System Development 2023/DAP_emulation/2021_surveys/PNP/06_imputation/backdata_output/PNP_2021_backdata_thousands.csv does not exist

To make it run you need to remove the postfix _thousand from the filename in dev_configs.

File: src/outlier_detection/auto_outliers.py (hunk 1 above)
Comment (AnneONS) [reply src/outlier_detection/auto_outliers.py:140]: Yes, the flag_outliers has been unit tested where the lower_clip is non zero. I think you're right, I'll add in a unit test for get_clip_bands for non-zero lower clip as well, and for when both are non-zero.

**Review by adriano-lopresti — APPROVED** [review]
